The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21877910124/job/63152019439) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/63152019439) [(plain)](https://github.com/rust-lang/rust/commit/071d0cd9f28d082e1cccd2c034b0e95d0b9c789b/checks/63152019439/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
REPOSITORY                                   TAG       IMAGE ID       CREATED       SIZE
ghcr.io/dependabot/dependabot-updater-core   latest    afc745c7535d   2 weeks ago   783MB
=> Removing docker images...
Deleted Images:
untagged: ghcr.io/dependabot/dependabot-updater-core:latest
untagged: ghcr.io/dependabot/dependabot-updater-core@sha256:faae3d3a1dedd24cde388bb506bbacc0f7ed1eae99ebac129af66acd8540c84a
deleted: sha256:afc745c7535da1bb12f92c273b9a7e9c52c3f12c5873714b2542da259c6d9769
deleted: sha256:64e147d5e54d9be8b8aa322e511cda02296eda4b8b8d063c6a314833aca50e29
deleted: sha256:5cba409bb463f4e7fa1a19f695450170422582c1bc7c0e934d893b4e5f558bc6
deleted: sha256:cddc6ebd344b0111eaab170ead1dfda24acdfe865ed8a12599a34d338fa8e28b
deleted: sha256:2412c3f334d79134573cd45e657fb6cc0abd75bef3881458b0d498d936545c8d
---
test [ui] tests/ui/abi/abi-sysv64-register-usage.rs ... ignored, ignored when the architecture is aarch64
test [ui] tests/ui/abi/abi-typo-unstable.rs#feature_disabled ... ok
test [ui] tests/ui/abi/abi-typo-unstable.rs#feature_enabled ... ok
test [ui] tests/ui/abi/anon-extern-mod.rs ... ok
test [ui] tests/ui/abi/avr-sram.rs#disable_sram ... ok
test [ui] tests/ui/abi/arm-unadjusted-intrinsic.rs#aarch64 ... ok
test [ui] tests/ui/abi/avr-sram.rs#has_sram ... ok
test [ui] tests/ui/abi/avr-sram.rs#no_sram ... ok
test [ui] tests/ui/abi/arm-unadjusted-intrinsic.rs#arm ... ok
test [ui] tests/ui/abi/bad-custom.rs ... ok
test [ui] tests/ui/abi/c-stack-as-value.rs ... ok
test [ui] tests/ui/abi/c-zst.rs#aarch64-darwin ... ok
test [ui] tests/ui/abi/c-zst.rs#powerpc-linux ... ok
---
test [ui] tests/ui/asm/aarch64/type-check-2-2.rs ... ok
test [ui] tests/ui/asm/aarch64/type-check-3.rs ... ok
test [ui] tests/ui/asm/aarch64/type-check-2.rs ... ok
test [ui] tests/ui/asm/aarch64/type-f16.rs ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#r82 ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#hf ... ok
test [ui] tests/ui/asm/aarch64/sym.rs ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#sf ... ok
test [ui] tests/ui/asm/arm-low-dreg.rs ... ok
test [ui] tests/ui/asm/binary_asm_labels.rs ... ignored, only executed when the architecture is x86_64
test [ui] tests/ui/asm/asm-with-nested-closure.rs ... ok
test [ui] tests/ui/asm/bad-template.rs#x86_64 ... ok
test [ui] tests/ui/asm/cfg.rs#reva ... ignored, only executed when the architecture is x86_64
---
test [ui] tests/ui/const-generics/occurs-check/unused-substs-2.rs ... ok
test [ui] tests/ui/const-generics/occurs-check/unused-substs-3.rs ... ok
test [ui] tests/ui/const-generics/occurs-check/unused-substs-5.rs ... ok
test [ui] tests/ui/const-generics/occurs-check/unused-substs-4.rs ... ok
test [ui] tests/ui/const-generics/ogca/basic-fail.rs ... ok
test [ui] tests/ui/const-generics/ogca/basic.rs ... ok
test [ui] tests/ui/const-generics/ogca/rhs-but-not-root.rs ... ok
test [ui] tests/ui/const-generics/ogca/coherence-ambiguous.rs ... ok
test [ui] tests/ui/const-generics/opaque_types.rs ... ok
test [ui] tests/ui/const-generics/opaque_types2.rs ... ok
test [ui] tests/ui/const-generics/outer-lifetime-in-const-generic-default.rs ... ok
test [ui] tests/ui/const-generics/overlapping_impls.rs ... ok
test [ui] tests/ui/const-generics/params-in-ct-in-ty-param-lazy-norm.rs#full ... ok
---
test [ui] tests/ui/extern/issue-64655-extern-rust-must-allow-unwind.rs#thin1 ... ok
test [ui] tests/ui/extern/issue-64655-extern-rust-must-allow-unwind.rs#fat3 ... ok
test [ui] tests/ui/extern/issue-80074.rs ... ok
test [ui] tests/ui/extern/issue-95829.rs ... ok
test [ui] tests/ui/extern/lgamma-linkage.rs ... ok
test [ui] tests/ui/extern/not-in-block.rs ... ok
test [ui] tests/ui/extern/no-mangle-associated-fn.rs ... ok
test [ui] tests/ui/extern/unsized-extern-derefmove.rs ... ok
test [ui] tests/ui/feature-gates/allow-features-empty.rs ... ok
test [ui] tests/ui/feature-gates/allow-features.rs ... ok
---
test [ui] tests/ui/feature-gates/feature-gate-macro-derive.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-macro-metavar-expr-concat.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-marker_trait_attr.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-may-dangle.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-mgca-type-const-syntax.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-min-generic-const-args.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-movrs_target_feature.rs ... ignored, only executed when the architecture is x86_64
test [ui] tests/ui/feature-gates/feature-gate-more-maybe-bounds.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-min_const_fn.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-naked_functions_rustic_abi.rs ... ignored, only executed when the architecture is x86_64
---
test [ui] tests/ui/imports/ambiguous-9.rs ... ok
test [ui] tests/ui/imports/ambiguous-import-visibility-module.rs ... ok
test [ui] tests/ui/imports/ambiguous-8.rs ... ok
test [ui] tests/ui/imports/ambiguous-glob-vs-expanded-extern.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-glob-vs-multiouter.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-globvsglob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-no-implicit-prelude.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-core-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-std-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-pick-core.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-pick-std.rs ... ok
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const/suggest-contraining-assoc-type-because-of-assoc-const.stderr`
diff of stderr:

6    |
7    = note: expected associated type `<C as O>::M`
8                          found type `u8`
-    = note: the associated type `<C as O>::M` is defined as `u8` in the implementation, but the generic bound `C` hides this definition
+    = note: the associated type `<C as O>::M` is defined as `u8` in the implementation, but the where-bound `C` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
10 help: consider constraining the associated type `<C as O>::M` to `u8`
11    |
12 LL | impl<C: O<M = u8>> U<C> for u16 {


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const.rs:13:21
   |
LL |     const N: C::M = 4u8; //~ ERROR mismatched types
   |                     ^^^ expected associated type, found `u8`
   |
   = note: expected associated type `<C as O>::M`
                         found type `u8`
   = note: the associated type `<C as O>::M` is defined as `u8` in the implementation, but the where-bound `C` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
help: consider constraining the associated type `<C as O>::M` to `u8`
   |
LL | impl<C: O<M = u8>> U<C> for u16 {
   |          ++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/associated-types/defaults-specialization/defaults-specialization.stderr`
diff of stderr:

75                          found type `u8`
76    = help: consider constraining the associated type `<A2<T> as Tr>::Ty` to `u8` or calling a method that returns `<A2<T> as Tr>::Ty`
77    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<A2<T> as Tr>::Ty` is defined as `u8` in the implementation, but the generic bound `A2<T>` hides this definition
+    = note: the associated type `<A2<T> as Tr>::Ty` is defined as `u8` in the implementation, but the where-bound `A2<T>` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
79 
80 error[E0308]: mismatched types
81   --> $DIR/defaults-specialization.rs:44:29

90    |
91    = note: expected associated type `<B2<T> as Tr>::Ty`
92                          found type `bool`
-    = note: the associated type `<B2<T> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B2<T>` hides this definition
+    = note: the associated type `<B2<T> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<T>` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
94 
95 error[E0308]: mismatched types
96   --> $DIR/defaults-specialization.rs:87:32

123    |
124 LL |     fn make() -> Self::Ty {
125    |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
-    = note: the associated type `<B<()> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B<()>` hides this definition
+    = note: the associated type `<B<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B<()>` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
127 
128 error[E0308]: mismatched types
129   --> $DIR/defaults-specialization.rs:89:33

156    |
157 LL |     fn make() -> Self::Ty {
158    |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
-    = note: the associated type `<B2<()> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B2<()>` hides this definition
+    = note: the associated type `<B2<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<()>` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
160 
161 error: aborting due to 9 previous errors; 1 warning emitted
162 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args associated-types/defaults-specialization.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-types/defaults-specialization.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/associated-types/defaults-specialization" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:3:38
   |
---
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:18
   |
LL |     fn make() -> Self::Ty {
   |                  ^^^^^^^^
   = note: expected signature `fn() -> <A<T> as Tr>::Ty`
              found signature `fn() -> u8`
help: change the output type to match the trait
   |
LL -     fn make() -> u8 { 0 }
LL +     fn make() -> <A<T> as Tr>::Ty { 0 }
   |

error[E0053]: method `make` has an incompatible type for trait
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:35:18
   |
---
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:18
   |
LL |     fn make() -> Self::Ty {
   |                  ^^^^^^^^
   = note: expected signature `fn() -> <B<T> as Tr>::Ty`
              found signature `fn() -> bool`
help: change the output type to match the trait
   |
LL -     fn make() -> bool { true }
LL +     fn make() -> <B<T> as Tr>::Ty { true }
   |

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:10:9
   |
---
   |
LL |     fn make() -> Self::Ty { 0u8 }
   |                  --------   ^^^ expected associated type, found `u8`
   |                  |
   |                  expected `<A2<T> as Tr>::Ty` because of return type
   |
   = note: expected associated type `<A2<T> as Tr>::Ty`
                         found type `u8`
   = help: consider constraining the associated type `<A2<T> as Tr>::Ty` to `u8` or calling a method that returns `<A2<T> as Tr>::Ty`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<A2<T> as Tr>::Ty` is defined as `u8` in the implementation, but the where-bound `A2<T>` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:44:29
   |
LL |     default type Ty = bool;
   |     --------------- associated type is `default` and may be overridden
LL |
LL |     fn make() -> Self::Ty { true }
   |                  --------   ^^^^ expected associated type, found `bool`
   |                  |
   |                  expected `<B2<T> as Tr>::Ty` because of return type
   |
   = note: expected associated type `<B2<T> as Tr>::Ty`
                         found type `bool`
   = note: the associated type `<B2<T> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<T>` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:87:32
   |
LL |     let _: <B<()> as Tr>::Ty = 0u8;   //~ error: mismatched types
   |            -----------------   ^^^ expected associated type, found `u8`
   |            |
   |            expected due to this
   |
   = note: expected associated type `<B<()> as Tr>::Ty`
                         found type `u8`
help: a method is available that returns `<B<()> as Tr>::Ty`
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:5
   |
LL |     fn make() -> Self::Ty {
   |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:88:32
   |
LL |     let _: <B<()> as Tr>::Ty = true;  //~ error: mismatched types
   |            -----------------   ^^^^ expected associated type, found `bool`
   |            |
   |            expected due to this
   |
   = note: expected associated type `<B<()> as Tr>::Ty`
                         found type `bool`
help: a method is available that returns `<B<()> as Tr>::Ty`
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:5
   |
LL |     fn make() -> Self::Ty {
   |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
   = note: the associated type `<B<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B<()>` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:89:33
   |
LL |     let _: <B2<()> as Tr>::Ty = 0u8;  //~ error: mismatched types
   |            ------------------   ^^^ expected associated type, found `u8`
   |            |
   |            expected due to this
   |
   = note: expected associated type `<B2<()> as Tr>::Ty`
                         found type `u8`
help: a method is available that returns `<B2<()> as Tr>::Ty`
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:5
   |
LL |     fn make() -> Self::Ty {
   |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:90:33
   |
LL |     let _: <B2<()> as Tr>::Ty = true; //~ error: mismatched types
   |            ------------------   ^^^^ expected associated type, found `bool`
   |            |
   |            expected due to this
   |
   = note: expected associated type `<B2<()> as Tr>::Ty`
                         found type `bool`
help: a method is available that returns `<B2<()> as Tr>::Ty`
  --> /checkout/tests/ui/associated-types/defaults-specialization.rs:9:5
   |
LL |     fn make() -> Self::Ty {
   |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
   = note: the associated type `<B2<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<()>` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error: aborting due to 9 previous errors; 1 warning emitted

Some errors have detailed explanations: E0053, E0308.
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-projection.current/specialization-default-projection.current.stderr`
diff of stderr:

21                     found unit type `()`
22    = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
23    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
25 
26 error[E0308]: mismatched types
27   --> $DIR/specialization-default-projection.rs:32:5

38            found associated type `<() as Foo>::Assoc`
39    = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
40    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `()` hides this definition
+    = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
42 
43 error: aborting due to 2 previous errors; 1 warning emitted
44 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args specialization/specialization-default-projection.rs`

error in revision `current`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/specialization/specialization-default-projection.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--cfg" "current" "--check-cfg" "cfg(test,FALSE,current,next)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-projection.current" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:5:12
   |
LL | #![feature(specialization)] //~ WARN the feature `specialization` is incomplete
   |            ^^^^^^^^^^^^^^
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:25:5
   |
LL | fn generic<T>() -> <T as Foo>::Assoc {
   |                    ----------------- expected `<T as Foo>::Assoc` because of return type
...
LL |     () //~ ERROR mismatched types
   |     ^^ expected associated type, found `()`
   |
   = note: expected associated type `<T as Foo>::Assoc`
                    found unit type `()`
   = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:32:5
   |
LL | fn monomorphic() -> () {
   |                     -- expected `()` because of return type
...
LL |     generic::<()>() //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^- help: consider using a semicolon here: `;`
   |     |
   |     expected `()`, found associated type
   |
   = note:    expected unit type `()`
           found associated type `<() as Foo>::Assoc`
   = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-projection.next/specialization-default-projection.next.stderr`
diff of stderr:

21                     found unit type `()`
22    = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
23    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
25 
26 error[E0308]: mismatched types
27   --> $DIR/specialization-default-projection.rs:32:5

38            found associated type `<() as Foo>::Assoc`
39    = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
40    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `()` hides this definition
+    = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
42 
43 error: aborting due to 2 previous errors; 1 warning emitted
44 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args specialization/specialization-default-projection.rs`

error in revision `next`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/specialization/specialization-default-projection.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--cfg" "next" "--check-cfg" "cfg(test,FALSE,current,next)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-projection.next" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Znext-solver"
stdout: none
--- stderr -------------------------------
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:5:12
   |
LL | #![feature(specialization)] //~ WARN the feature `specialization` is incomplete
   |            ^^^^^^^^^^^^^^
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:25:5
   |
LL | fn generic<T>() -> <T as Foo>::Assoc {
   |                    ----------------- expected `<T as Foo>::Assoc` because of return type
...
LL |     () //~ ERROR mismatched types
   |     ^^ types differ
   |
   = note: expected associated type `<T as Foo>::Assoc`
                    found unit type `()`
   = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-projection.rs:32:5
   |
LL | fn monomorphic() -> () {
   |                     -- expected `()` because of return type
...
LL |     generic::<()>() //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^- help: consider using a semicolon here: `;`
   |     |
   |     types differ
   |
   = note:    expected unit type `()`
           found associated type `<() as Foo>::Assoc`
   = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-types.current/specialization-default-types.current.stderr`
diff of stderr:

20    |
21    = note: expected associated type `<T as Example>::Output`
22                        found struct `Box<T>`
-    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
24 
25 error[E0308]: mismatched types
26   --> $DIR/specialization-default-types.rs:29:5

34            found associated type `<T as Example>::Output`
35    = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
36    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
38 
39 error: aborting due to 2 previous errors; 1 warning emitted
40 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args specialization/specialization-default-types.rs`

error in revision `current`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/specialization/specialization-default-types.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--cfg" "current" "--check-cfg" "cfg(test,FALSE,current,next)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-types.current" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:9:12
   |
LL | #![feature(specialization)] //~ WARN the feature `specialization` is incomplete
   |            ^^^^^^^^^^^^^^
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:19:9
   |
LL |     default type Output = Box<T>;
   |     ------------------- associated type is `default` and may be overridden
LL |     default fn generate(self) -> Self::Output {
   |                                  ------------ expected `<T as Example>::Output` because of return type
LL |         Box::new(self) //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^^ expected associated type, found `Box<T>`
   |
   = note: expected associated type `<T as Example>::Output`
                       found struct `Box<T>`
   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:29:5
   |
LL | fn trouble<T>(t: T) -> Box<T> {
   |                        ------ expected `Box<T>` because of return type
LL |     Example::generate(t) //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^^^^^^ expected `Box<T>`, found associated type
   |
   = note:       expected struct `Box<T>`
           found associated type `<T as Example>::Output`
   = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
---
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-types.next/specialization-default-types.next.stderr`
diff of stderr:

20    |
21    = note: expected associated type `<T as Example>::Output`
22                        found struct `Box<T>`
-    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
24 
25 error[E0308]: mismatched types
26   --> $DIR/specialization-default-types.rs:29:5

34            found associated type `<T as Example>::Output`
35    = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
36    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+    = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+            see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
38 
39 error: aborting due to 2 previous errors; 1 warning emitted
40 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args specialization/specialization-default-types.rs`

error in revision `next`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/specialization/specialization-default-types.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--cfg" "next" "--check-cfg" "cfg(test,FALSE,current,next)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/specialization/specialization-default-types.next" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Znext-solver"
stdout: none
--- stderr -------------------------------
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:9:12
   |
LL | #![feature(specialization)] //~ WARN the feature `specialization` is incomplete
   |            ^^^^^^^^^^^^^^
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:19:9
   |
LL |     default type Output = Box<T>;
   |     ------------------- associated type is `default` and may be overridden
LL |     default fn generate(self) -> Self::Output {
   |                                  ------------ expected `<T as Example>::Output` because of return type
LL |         Box::new(self) //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^^ types differ
   |
   = note: expected associated type `<T as Example>::Output`
                       found struct `Box<T>`
   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/specialization/specialization-default-types.rs:29:5
   |
LL | fn trouble<T>(t: T) -> Box<T> {
   |                        ------ expected `Box<T>` because of return type
LL |     Example::generate(t) //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^^^^^^ types differ
   |
   = note:       expected struct `Box<T>`
           found associated type `<T as Example>::Output`
   = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
```

</details>

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
@@ -58,6 +58,7 @@ use rustc_hir::def_id::DefId;
 use rustc_hir::intravisit::Visitor;
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::{self as hir};
+use rustc_infer::infer::DefineOpaqueTypes;
 use rustc_macros::extension;
 use rustc_middle::bug;
 use rustc_middle::dep_graph::DepContext;
@@ -278,6 +279,21 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
 
             let is_shadowed = self.infcx.probe(|_| {
                 let impl_substs = self.infcx.fresh_args_for_item(DUMMY_SP, impl_def_id);
+                let impl_trait_ref =
+                    tcx.impl_trait_ref(impl_def_id).instantiate(tcx, impl_substs);
+
+                let expected_trait_ref = alias.trait_ref(tcx);
+
+                if let Err(_) = self.infcx.at(&ObligationCause::dummy(), param_env).eq(
+                    DefineOpaqueTypes::No,
+                    expected_trait_ref,
+                    impl_trait_ref,
+                ) {
+                    return false;
+                }
+
+                let trait_def_id = alias.trait_def_id(tcx);
+                let rebased_args = alias.args.rebase_onto(tcx, trait_def_id, impl_substs);
 
                 let leaf_def = match specialization_graph::assoc_def(tcx, impl_def_id, alias.def_id)
                 {
@@ -286,7 +302,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                 };
 
                 let impl_item_def_id = leaf_def.item.def_id;
-                let impl_assoc_ty = tcx.type_of(impl_item_def_id).instantiate(tcx, impl_substs);
+                let impl_assoc_ty = tcx.type_of(impl_item_def_id).instantiate(tcx, rebased_args);
 
                 self.infcx.can_eq(param_env, impl_assoc_ty, concrete)
             });

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
@@ -279,8 +279,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
 
             let is_shadowed = self.infcx.probe(|_| {
                 let impl_substs = self.infcx.fresh_args_for_item(DUMMY_SP, impl_def_id);
-                let impl_trait_ref =
-                    tcx.impl_trait_ref(impl_def_id).instantiate(tcx, impl_substs);
+                let impl_trait_ref = tcx.impl_trait_ref(impl_def_id).instantiate(tcx, impl_substs);
 
                 let expected_trait_ref = alias.trait_ref(tcx);
 

--- a/tests/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const.stderr
+++ b/tests/ui/associated-type-bounds/suggest-contraining-assoc-type-because-of-assoc-const.stderr
@@ -6,7 +6,8 @@ LL |     const N: C::M = 4u8;
    |
    = note: expected associated type `<C as O>::M`
                          found type `u8`
-   = note: the associated type `<C as O>::M` is defined as `u8` in the implementation, but the generic bound `C` hides this definition
+   = note: the associated type `<C as O>::M` is defined as `u8` in the implementation, but the where-bound `C` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 help: consider constraining the associated type `<C as O>::M` to `u8`
    |
 LL | impl<C: O<M = u8>> U<C> for u16 {

--- a/tests/ui/associated-types/defaults-specialization.stderr
+++ b/tests/ui/associated-types/defaults-specialization.stderr
@@ -75,7 +75,8 @@ LL |     fn make() -> Self::Ty { 0u8 }
                          found type `u8`
    = help: consider constraining the associated type `<A2<T> as Tr>::Ty` to `u8` or calling a method that returns `<A2<T> as Tr>::Ty`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<A2<T> as Tr>::Ty` is defined as `u8` in the implementation, but the generic bound `A2<T>` hides this definition
+   = note: the associated type `<A2<T> as Tr>::Ty` is defined as `u8` in the implementation, but the where-bound `A2<T>` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/defaults-specialization.rs:44:29
@@ -90,7 +91,8 @@ LL |     fn make() -> Self::Ty { true }
    |
    = note: expected associated type `<B2<T> as Tr>::Ty`
                          found type `bool`
-   = note: the associated type `<B2<T> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B2<T>` hides this definition
+   = note: the associated type `<B2<T> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<T>` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/defaults-specialization.rs:87:32
@@ -123,7 +125,8 @@ help: a method is available that returns `<B<()> as Tr>::Ty`
    |
 LL |     fn make() -> Self::Ty {
    |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
-   = note: the associated type `<B<()> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B<()>` hides this definition
+   = note: the associated type `<B<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B<()>` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/defaults-specialization.rs:89:33
@@ -156,7 +159,8 @@ help: a method is available that returns `<B2<()> as Tr>::Ty`
    |
 LL |     fn make() -> Self::Ty {
    |     ^^^^^^^^^^^^^^^^^^^^^ consider calling `Tr::make`
-   = note: the associated type `<B2<()> as Tr>::Ty` is defined as `bool` in the implementation, but the generic bound `B2<()>` hides this definition
+   = note: the associated type `<B2<()> as Tr>::Ty` is defined as `bool` in the implementation, but the where-bound `B2<()>` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error: aborting due to 9 previous errors; 1 warning emitted
 

--- a/tests/ui/specialization/specialization-default-projection.current.stderr
+++ b/tests/ui/specialization/specialization-default-projection.current.stderr
@@ -21,7 +21,8 @@ LL |     ()
                     found unit type `()`
    = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/specialization-default-projection.rs:32:5
@@ -38,7 +39,8 @@ LL |     generic::<()>()
            found associated type `<() as Foo>::Assoc`
    = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `()` hides this definition
+   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error: aborting due to 2 previous errors; 1 warning emitted
 

--- a/tests/ui/specialization/specialization-default-projection.next.stderr
+++ b/tests/ui/specialization/specialization-default-projection.next.stderr
@@ -21,7 +21,8 @@ LL |     ()
                     found unit type `()`
    = help: consider constraining the associated type `<T as Foo>::Assoc` to `()` or calling a method that returns `<T as Foo>::Assoc`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/specialization-default-projection.rs:32:5
@@ -38,7 +39,8 @@ LL |     generic::<()>()
            found associated type `<() as Foo>::Assoc`
    = help: consider constraining the associated type `<() as Foo>::Assoc` to `()`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the generic bound `()` hides this definition
+   = note: the associated type `<() as Foo>::Assoc` is defined as `()` in the implementation, but the where-bound `()` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error: aborting due to 2 previous errors; 1 warning emitted
 

--- a/tests/ui/specialization/specialization-default-types.current.stderr
+++ b/tests/ui/specialization/specialization-default-types.current.stderr
@@ -20,7 +20,8 @@ LL |         Box::new(self)
    |
    = note: expected associated type `<T as Example>::Output`
                        found struct `Box<T>`
-   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/specialization-default-types.rs:29:5
@@ -34,7 +35,8 @@ LL |     Example::generate(t)
            found associated type `<T as Example>::Output`
    = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error: aborting due to 2 previous errors; 1 warning emitted
 

--- a/tests/ui/specialization/specialization-default-types.next.stderr
+++ b/tests/ui/specialization/specialization-default-types.next.stderr
@@ -20,7 +20,8 @@ LL |         Box::new(self)
    |
    = note: expected associated type `<T as Example>::Output`
                        found struct `Box<T>`
-   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error[E0308]: mismatched types
   --> $DIR/specialization-default-types.rs:29:5
@@ -34,7 +35,8 @@ LL |     Example::generate(t)
            found associated type `<T as Example>::Output`
    = help: consider constraining the associated type `<T as Example>::Output` to `Box<T>`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
-   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the generic bound `T` hides this definition
+   = note: the associated type `<T as Example>::Output` is defined as `Box<T>` in the implementation, but the where-bound `T` shadows this definition
+           see issue #152409 <https://github.com/rust-lang/rust/issues/152409> for more information
 
 error: aborting due to 2 previous errors; 1 warning emitted
 

